pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,760 @@
+
+	log := clog.FromContext(ctx)
+	stdout, stderr := logwriter.New(log.Info), logwriter.New(log.Warn)
+	defer stdout.Close()
+	defer stderr.Close()
+
+	// default to root user but if a different user is specified
+	// we will use the embedded build:1000:1000 user
+	user := "root"
+	if cfg.RunAs != "" {
+		user = "build"
+	}
+
+	err := sendSSHCommand(ctx,
+		user,
+		"localhost",
+		cfg.SSHPort,
+		cfg,
+		envOverride,
+		nil,
+		stderr,
+		stdout,
+		args,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bw *qemu) Debug(ctx context.Context, cfg *Config, envOverride map[string]string, args ...string) error {
+	clog.InfoContextf(ctx, "debugging command %s", strings.Join(args, " "))
+
+	// default to root user but if a different user is specified
+	// we will use the embedded build:1000:1000 user
+	user := "root"
+	if cfg.RunAs != "" {
+		user = "build"
+	}
+
+	err := sendSSHCommand(ctx,
+		user,
+		"localhost",
+		cfg.SSHPort,
+		cfg,
+		envOverride,
+		os.Stdin,
+		os.Stderr,
+		os.Stdout,
+		args,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TestUsability determines if the Qemu runner can be used
+// as a microvm runner.
+func (bw *qemu) TestUsability(ctx context.Context) bool {
+	log := clog.FromContext(ctx)
+
+	arch := apko_types.Architecture(runtime.GOARCH)
+	if _, err := exec.LookPath(fmt.Sprintf("qemu-system-%s", arch.ToAPK())); err != nil {
+		log.Warnf("cannot use qemu for microvms: qemu-system-%s not found on $PATH", arch.ToAPK())
+		return false
+	}
+
+	return true
+}
+
+// OCIImageLoader used to load OCI images in, if needed. qemu does not need it.
+func (bw *qemu) OCIImageLoader() Loader {
+	return &qemuOCILoader{}
+}
+
+// TempDir returns the base for temporary directory. For qemu, this is empty.
+func (bw *qemu) TempDir() string {
+	return ""
+}
+
+// StartPod starts a pod if necessary.
+func (bw *qemu) StartPod(ctx context.Context, cfg *Config) error {
+	ctx, span := otel.Tracer("melange").Start(ctx, "qemu.StartPod")
+	defer span.End()
+
+	sshPort, err := randpomPortN()
+	if err != nil {
+		return err
+	}
+
+	cfg.SSHPort = strconv.Itoa(sshPort)
+
+	return createMicroVM(ctx, cfg)
+}
+
+// TerminatePod terminates a pod if necessary.  Not implemented
+// for Qemu runners.
+func (bw *qemu) TerminatePod(ctx context.Context, cfg *Config) error {
+	ctx, span := otel.Tracer("melange").Start(ctx, "qemu.TerminatePod")
+	defer span.End()
+	defer os.Remove(cfg.ImgRef)
+
+	clog.FromContext(ctx).Info("qemu: sending shutdown signal")
+	err := sendSSHCommand(ctx,
+		"root",
+		"localhost",
+		cfg.SSHPort,
+		cfg,
+		nil,
+		nil,
+		nil,
+		nil,
+		[]string{"echo s > /proc/sysrq-trigger && echo o > /proc/sysrq-trigger&"},
+	)
+	if err != nil {
+		return err
+	}
+
+	if cfg.Disk != "" {
+		err = os.Remove(cfg.Disk)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// WorkspaceTar implements Runner
+func (bw *qemu) WorkspaceTar(ctx context.Context, cfg *Config) (io.ReadCloser, error) {
```

Comment:
this never returns anything other than nil?
i'm guessing this is just in progress.

Reply:
for some runners (see also bwrap) this is a noop
in this case it's not properly a noop, but does not return a tar indeed